wpilib/wpilib/pidcontroller.py

```diff
@@ -519,41 +507,91 @@ def reset(self):
+
     def getSmartDashboardType(self):
         return "PIDController"
 
-    def valueChanged(self, table, key, value, isNew):
-        if key == "p" or key == "i" or key == "d" or key == "f":
-            Kp = table.getNumber("p", 0.0)
-            Ki = table.getNumber("i", 0.0)
-            Kd = table.getNumber("d", 0.0)
-            Kf = table.getNumber("f", 0.0)
-            if (self.getP() != Kp or self.getI() != Ki or self.getD() != Kd or
-                self.getF() != Kf):
-                self.setPID(Kp, Ki, Kd, Kf)
-        elif key == "setpoint":
-            if self.getSetpoint() != float(value):
-                self.setSetpoint(float(value))
-        elif key == "enabled":
-            if self.isEnabled() != bool(value):
-                if bool(value):
-                    self.enable()
-                else:
-                    self.disable()
+    def pChanged(self, entry, key, value, param):
+        with self.mutex:
+            self.P = value
+
+    def iChanged(self, entry, key, value, param):
+        with self.mutex:
+            self.I = value
+
+    def dChanged(self, entry, key, value, param):
+        with self.mutex:
+            self.D = value
+
+    def fChanged(self, entry, key, value, param):
+        with self.mutex:
+            self.F = value
+
+    def setpointChanged(self, entry, key, value, param):
+        if self.getSetpoint() != value:
+            self.setSetpoint(value)
+
+    def enabledChanged(self, entry, key, value, param):
+        if self.isEnabled() != value:
+            if value:
+                self.enable()
+            else:
+                self.disable()
 
     def initTable(self, table):
-        oldtable = self.getTable()
-        if oldtable is not None:
-            oldtable.removeTableListener(self.valueChanged)
-        self.table = table
+        self.removeListeners()
         if table is not None:
-            table.putNumber("p", self.getP())
-            table.putNumber("i", self.getI())
-            table.putNumber("d", self.getD())
-            table.putNumber("f", self.getF())
-            table.putNumber("setpoint", self.getSetpoint())
-            table.putBoolean("enabled", self.isEnabled())
-            table.addTableListener(self.valueChanged, False)
+            self.pEntry = table.getEntry("p")
+            self.pEntry.setDouble(self.getP())
+            self.iEntry = table.getEntry("i")
+            self.iEntry.setDouble(self.getI())
+            self.dEntry = table.getEntry("d")
+            self.dEntry.setDouble(self.getD())
+            self.fEntry = table.getEntry("f")
+            self.fEntry.setDouble(self.getF())
+            self.setpointEntry = table.getEntry("setpoint")
+            self.setpointEntry.setDouble(self.getSetpoint())
+            self.enabledEntry = table.getEntry("enabled")
+            self.enabledEntry.setBoolean(self.isEnabled())
+
+            self.pListener = self.pEntry.addListener(
```

Comment:
Ouch. Adding all of these listeners is going to end up with a ton of overhead. I know you didn't do this so no fix required here -- but we need to file a bug upstream.

Reply:
Well.. on second thought, if that's only active during LiveWindow mode, then I don't really care. Unless it's active all the time?